src/core/symbology/qgsgraduatedsymbolrenderer.cpp

```diff
@@ -525,29 +531,83 @@ QgsSymbolList QgsGraduatedSymbolRenderer::symbols( QgsRenderContext &context ) c
   return lst;
 }
 
-static QList<double> _calcEqualIntervalBreaks( double minimum, double maximum, int classes )
+static void _makeBreaksSymmetric( QList<double> &breaks, double symmetryPoint, bool astride )
 {
+  // remove the breaks that are above the existing opposite sign classes
+  // to keep colors symmetrically balanced around symmetryPoint
+  // if astride is true, remove the symmetryPoint break so that
+  // the 2 classes form only one
 
-  // Equal interval algorithm
-  //
-  // Returns breaks based on dividing the range ('minimum' to 'maximum')
-  // into 'classes' parts.
-
-  double step = ( maximum - minimum ) / classes;
+  if ( breaks.size() > 1 ) //to avoid crash when only 1 class
+  {
+    std::sort( breaks.begin(), breaks.end() );
+    // breaks contain the maximum of the distrib but not the minimum
+    double distBelowSymmetricValue = std::abs( breaks[0] - symmetryPoint );
+    double distAboveSymmetricValue = std::abs( breaks[ breaks.size() - 2 ] - symmetryPoint ) ;
+    double absMin = std::min( std::abs( distAboveSymmetricValue ), std::abs( distBelowSymmetricValue ) );
+    // make symmetric
+    for ( int i = 0; i <= breaks.size() - 2; ++i )
+    {
+      if ( std::abs( breaks.at( i ) - symmetryPoint ) > absMin )
+      {
+        breaks.removeAt( i );
+        --i;
+      }
+    }
+    // remove symmetry point
+    if ( true == astride ) // && breaks.indexOf( symmetryPoint ) != -1) // if symmetryPoint is found
+    {
+      breaks.removeAt( breaks.indexOf( symmetryPoint ) );
+    }
+  }
+}
 
+static QList<double> _calcEqualIntervalBreaks( double minimum, double maximum, int classes, bool useSymmetricMode, double symmetryPoint, bool astride )
+{
+  // Equal interval algorithm
+  // Returns breaks based on dividing the range ('minimum' to 'maximum') into 'classes' parts.
   QList<double> breaks;
-  double value = minimum;
-  breaks.reserve( classes );
-  for ( int i = 0; i < classes; i++ )
+  if ( false == useSymmetricMode ) // nomal mode
   {
-    value += step;
-    breaks.append( value );
+    double step = ( maximum - minimum ) / classes;
+
+    double value = minimum;
+    breaks.reserve( classes );
+    for ( int i = 0; i < classes; i++ )
+    {
+      value += step;
+      breaks.append( value );
+    }
+    // floating point arithmetics is not precise:
+    // set the last break to be exactly maximum so we do not miss it
+    breaks[classes - 1] = maximum;
   }
+  else if ( true == useSymmetricMode ) // symmetric mode
```

Comment:
This one too  (and the others I haven't flagged) - QGIS code convention is to test bools/pointers directly and not compare against true/false